File: BAEKJOON/2805.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm> 
using namespace std;

int N,M;

int main() {
    cin >> N >> M;
    
    vector<int> trees(N);

    int tree,result;
    

    result=0;

    for (int i = 0; i < N; i++) {
        cin >> trees[i];
    }

    sort(trees.begin(), trees.end());
    int start = 0;
    int end = trees.back();

    while (start <= end) {
        int mid = (start + end)/2;
        long long total = 0;
        for (int i= 0; i<N; i ++){
            if (trees[i] > mid){
                total+=trees[i]-mid;
            }
        }

        if (total >= M){
            result = mid;
            start = mid +1;
        }
        else
            end = mid -1;
    }

    cout << result;
}

File: BAEKJOON/2805.py
import sys
input = sys.stdin.readline

N, M = map(int, input().strip().split())

trees = list(map(int, input().strip().split()))

start, end = 0, max(trees)
result = 0

while start <= end:
    mid = (start + end) // 2  
    total = 0

    
    for tree in trees:
        if tree > mid:
            total += tree - mid
    
    if total >= M:
        result = mid  # 현재 mid 높이를 기록
        start = mid + 1  # 더 큰 높이를 시도
    else:
        end = mid - 1  # 더 낮은 높이를 시도

# 결과 출력
print(result)
